File: MATTHEW/leetcode/AIcoding/5.Qwen-3/rust_solutions/576-out-of-boundary-paths.rs
struct Solution;

impl Solution {
    pub fn find_paths(m: i32, n: i32, max_moves: i32, start_row: i32, start_col: i32) -> i32 {
        use std::collections::HashMap;

        type State = (i32, i32, i32);
        let mut memo = HashMap::new();

        fn dfs(
            m: i32,
            n: i32,
            moves_left: i32,
            row: i32,
            col: i32,
            memo: &mut HashMap<(i32, i32, i32), i32>,
        ) -> i32 {
            if row < 0 || row >= m || col < 0 || col >= n {
                return 1;
            }
            if moves_left == 0 {
                return 0;
            }
            let key = (row, col, moves_left);
            if let Some(&result) = memo.get(&key) {
                return result;
            }

            let directions = [(0, 1), (1, 0), (0, -1), (-1, 0)];
            let mut count = 0;
            for &(dr, dc) in &directions {
                count += dfs(
                    m,
                    n,
                    moves_left - 1,
                    row + dr,
                    col + dc,
                    memo,
                );
            }

            memo.insert(key, count);
            count
        }

        dfs(m, n, max_moves, start_row, start_col, &mut memo)
    }
}

File: leetcode/AIcoding/5.Qwen-3/java_solutions/576-out-of-boundary-paths.java
public class Solution {
    public int findPaths(int m, int n, int N, int x, int y) {
        int[][][] dp = new int[m][n][N + 1];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j][0] = 0;
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 || i == m - 1 || j == 0 || j == n - 1) {
                    dp[i][j][1] = 1;
                } else {
                    dp[i][j][1] = 0;
                }
            }
        }
        for (int step = 2; step <= N; step++) {
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    int count = 0;
                    if (i - 1 < 0) count++;
                    else count += dp[i - 1][j][step - 1];
                    if (i + 1 >= m) count++;
                    else count += dp[i + 1][j][step - 1];
                    if (j - 1 < 0) count++;
                    else count += dp[i][j - 1][step - 1];
                    if (j + 1 >= n) count++;
                    else count += dp[i][j + 1][step - 1];
                    dp[i][j][step] = count;
                }
            }
        }
        return dp[x][y][N];
    }
}